Migrate Employees list actions

sarjon opened the pull request:
<!--
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
 -->

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | This PR migrates actions for Employee list: "bulk delete", "bulk enable", "bulk disable", "delete" row action and "toggle" action in "Active" column.
| Type?         | refacto
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | https://github.com/PrestaShop/PrestaShop/issues/10502
| How to test?  | Access `admin-dev/index.php/configure/advanced/employees/` and check actions that are migrated. **NOTE: it does not hide actions for current employee yet.**
<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/11135)
<!-- Reviewable:end -->

sarjon committed:
use common search

sarjon committed:
start implementing employee toggle command

sarjon committed:
finish employee toggle command

sarjon committed:
add permissions and demo check

sarjon committed:
update assertion naming

sarjon committed:
wip on update employee status

sarjon committed:
add abstract employee update handler with assertions

sarjon committed:
update employee status in command handler

sarjon committed:
add employees status update handler

sarjon committed:
implement bulk status update

sarjon committed:
add employee delete command

sarjon committed:
handle warehouse manager deletion

sarjon committed:
add bulk delete of employees

sarjon committed:
added security and demo mode checks

sarjon committed:
fix handlers name

sarjon committed:
fix typo in flashing message

sarjon committed:
remove docs duplication

sarjon committed:
make BulkUpdateEmployeeStatusHandler class definition fit the line

sarjon committed:
add missing docblock

sarjon committed:
update employee route naming

sarjon committed:
apply cs fixer

mickaelandrieu reviewed (dismissed):
Would you mind to add tests?

I see so much complexity and layers introduced here, the gains were supposed to be easier to test the behavior but afaik no tests were added with CQRS implementation.

I think it's time to prove that CQRS is actually more than "one more" extra layer: don't you think?

tomas862 commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/EmployeeController.php:
```diff
@@ -67,59 +79,179 @@ public function indexAction(Request $request, EmployeeFilters $filters)
      *
-     * @AdminSecurity("is_granted(['read', 'update', 'create', 'delete'], request.get('_legacy_controller'))")
+     * @DemoRestricted(redirectRoute="admin_employees_index")
+     * @AdminSecurity("is_granted(['update', 'create', 'delete'], request.get('_legacy_controller'))")
```
```suggestion
     * @AdminSecurity(
     "is_granted(['update', 'create', 'delete'], request.get('_legacy_controller'))",
     redirectRoute="admin_employees_index"
     )
```

tomas862 commented on src/Adapter/Profile/Employee/CommandHandler/DeleteEmployeeHandler.php:
```diff
@@ -0,0 +1,53 @@
+use PrestaShop\PrestaShop\Core\Domain\Profile\Employee\CommandHandler\DeleteEmployeeHandlerInterface;
+
+/**
+ * Class DeleteEmployeeHandler.
```
Missing description

tomas862 commented on src/Adapter/Profile/Employee/CommandHandler/AbstractEmployeeHandler.php:
```diff
@@ -0,0 +1,114 @@
+use Warehouse;
+
+/**
+ * Class AbstractEmployeeStatusHandler.
```
Missing description

tomas862 commented on src/Adapter/Profile/Employee/CommandHandler/BulkDeleteEmployeeHandler.php:
```diff
@@ -0,0 +1,54 @@
+use PrestaShop\PrestaShop\Core\Domain\Profile\Employee\CommandHandler\BulkDeleteEmployeeHandlerInterface;
+
+/**
+ * Class BulkDeleteEmployeeHandler.
```
Missing description

sarjon commented:
> Would you mind to add tests?

yup, i would. :+1: 

but what would be the correct way to do it? i dont want to add slow tests, because each (probably?) command/handler would require setting up and reseting database after test case, right?

matks commented:
> > Would you mind to add tests?
> 
> yup, i would. 👍
> 
> but what would be the correct way to do it? i dont want to add slow tests, because each (probably?) command/handler would require setting up and reseting database after test case, right?

Very important topic 😄 let's find the answer here:
https://github.com/PrestaShop/PrestaShop/issues/11191

matks commented on src/Core/Domain/Profile/Employee/Exception/UndefinedEmployeeStatusException.php:
```diff
@@ -0,0 +1,34 @@
+/**
+ * Class UndefinedEmployeeAccountStatusException.
+ */
+class UndefinedEmployeeStatusException extends EmployeeException
```
That's a lot of domain Exceptions 🤔 maybe too much ?

sarjon replied:
i dont know.. i dont want to throw single generic exception (e.g. `EmployeeException`) for different errors, do you have other ideas? :thinking:

matks replied:
Sometimes, you use exception codes as php constants. Looks good IMO.

sarjon replied:
yup, when constraint exception happens. Since its pretty generic, we use codes to specify which constraint was violated. do you suggest having just one `EmployeeException` with a lot of codes?

matks replied:
No you're right 😄 thank you

matks commented on src/Adapter/Profile/Employee/CommandHandler/BulkDeleteEmployeeHandler.php:
```diff
@@ -0,0 +1,54 @@
+     */
+    public function handle(BulkDeleteEmployeeCommand $command)
+    {
+        foreach ($command->getEmployeeIds() as $employeeId) {
```
If one `delete()` fails (throws an exception), then the Command will halt halfway, leaving some of the Employees deleted and other not deleted. A try/catch here would prevent that. But you have to handle the return then 😄 what will you return ? An array of success/errors ? A multi-exception for each failure ?

sarjon replied:
lets throw exception (meaning that we will skip remaining employees) if it fails to delete employee, we have discussed this case with @eternoendless when doing first commands for SQL Manager. :)

matks commented on src/Adapter/Profile/Employee/CommandHandler/BulkUpdateEmployeeStatusHandler.php:
```diff
@@ -0,0 +1,54 @@
+     */
+    public function handle(BulkUpdateEmployeeStatusCommand $command)
+    {
+        foreach ($command->getEmployeeIds() as $employeeId) {
```
Same here, how do we deal with halfway failure

matks commented on src/Core/Domain/Profile/Employee/ValueObject/EmployeeStatus.php:
```diff
@@ -0,0 +1,93 @@
+    /**
+     * @param string $status
+     */
+    private function assertStatusExists($status)
```
👍

matks commented on src/Core/Domain/Profile/Employee/ValueObject/EmployeeStatus.php:
```diff
@@ -0,0 +1,93 @@
+/**
+ * Class EmployeeStatus.
+ */
+class EmployeeStatus
```
This class can get unit tests :)

sarjon replied:
done. :+1:

matks commented on src/Core/Domain/Profile/Employee/ValueObject/EmployeeStatus.php:
```diff
@@ -0,0 +1,93 @@
+    /**
+     * @internal
+     */
+    const AVAILABLE_STATUS = [
```
`AVAILABLE_STATUSES` ?

matks commented on src/Core/Domain/Profile/Employee/ValueObject/EmployeeId.php:
```diff
@@ -0,0 +1,73 @@
+/**
+ * Class EmployeeId.
+ */
+class EmployeeId
```
I know we started using more and more of them but I'm still not sure about it 🤔. Let's keep them for now but we need to make sure whether we use them before 1.7.6 release.

matks commented on src/PrestaShopBundle/Resources/config/routing/admin/configure/advanced_parameters/employee.yml:
```diff
@@ -9,12 +9,43 @@ admin_employees_search:
-    _controller: 'PrestaShopBundle:Admin\Configure\AdvancedParameters\Employee:search'
-    _legacy_controller: AdminEmployees
+    _controller: 'PrestaShopBundle:Admin\Common:searchGrid'
+    gridDefinitionFactoryService: prestashop.core.grid.definition.factory.employee
```
Why this ?

sarjon replied:
its a common search action, we need to provide it with grid definition factory so it can create filters form.

matks replied:
Maybe we should write doc about, wdyt ?

sarjon replied:
probably, but its not related to this PR.

sarjon committed:
handle employee deletion exception

sarjon committed:
fix naming

sarjon committed:
add test for EmployeeStatus

matks commented:
@sarjon OK, only need to discuss further
https://github.com/PrestaShop/PrestaShop/pull/11135#discussion_r231481089
and
https://github.com/PrestaShop/PrestaShop/pull/11135#discussion_r231476326

sarjon commented:
anything else needed for this PR?

matks commented:
@sarjon you're right, let's go for QA then 😄

marionf commented:
@sarjon 

I have this issue when I try to login in BO

![capture d ecran_784](https://user-images.githubusercontent.com/13449658/50148232-3d77a500-02b8-11e9-9fe9-97e77d5d570b.png)

matks commented:
@marionf I fixed the branch and I checked: it works. So I put back label "waiting for QA" 😉

marionf commented:
@sarjon 

1) The checkbox and the delete button shouldn't be displayed for the super admin created during the install

![capture d ecran_861](https://user-images.githubusercontent.com/13449658/50761058-b50e7380-1269-11e9-80ad-59374ad4a59d.png)

![capture d ecran_862](https://user-images.githubusercontent.com/13449658/50761098-d0797e80-1269-11e9-835b-f92a25fc06ec.png)


2) The actions label is missing and the search/reset button should be aligned on the right

![capture d ecran_863](https://user-images.githubusercontent.com/13449658/50761240-29491700-126a-11e9-95c7-20569cfe2c3c.png)


3) Wheel button should be in the header

![capture d ecran_865](https://user-images.githubusercontent.com/13449658/50761969-fbfd6880-126b-11e9-9318-27d8e90c942f.png)


4) I don't see the button to add a new employee

![capture d ecran_866](https://user-images.githubusercontent.com/13449658/50762189-8f369e00-126c-11e9-9234-af28e85569d8.png)


5) Elements of the pagination are not aligned

![capture d ecran_867](https://user-images.githubusercontent.com/13449658/50762496-66fb6f00-126d-11e9-858d-7292dad6d918.png)

![50214246-0ff62e80-0380-11e9-8d41-7af751038e15 1](https://user-images.githubusercontent.com/13449658/50762682-f3a62d00-126d-11e9-9539-4db2f3530ec8.png)


6) The edit button is not working

![capture d ecran_868](https://user-images.githubusercontent.com/13449658/50762779-4849a800-126e-11e9-95b0-6a9c5b5b1cf3.png)

eternoendless commented:
> 3. Wheel button should be in the header
> ![Screencap](https://user-images.githubusercontent.com/13449658/50761969-fbfd6880-126b-11e9-9318-27d8e90c942f.png)

Are you sure about this? The [product catalog listing](https://projects.invisionapp.com/d/main#/console/13800704/287602258/preview) doesn't have a header, maybe this should be normalized one way or the other /cc @TristanLDD

marionf commented:
@eternoendless it has been done for all migrated pages
See #11095 and #11080

eternoendless commented:
Okay then

TristanLDD commented:
@eternoendless  Indeed, the product catalog listing doesn't have a header, as a consequence the wheel is placed on top right and this page represents an exception. But it was agreed that the actions which refer to a specific container are placed in the container's header, so the behavior is already normalized for the majority of cases ;)

sarjon committed:
fix syntax error

sarjon committed:
align right search & reset buttons in grid

sarjon committed:
add name to actions column

sarjon committed:
add Add new Employee button in list

sarjon committed:
add placeholders for grid fitlers

sarjon committed:
move employee options help text below input

sarjon commented:
@marionf regarding your feedback:

1. This is part of another PR since this functionality needs to be improved.
2. Done.
3. Done.
4. I've added "Add new Employee" button, but keep in mind that Add/Edit is not migrated for Employee, so both Add and Edit buttons are not working.
5. Done
6. Related to 4.

Additional changes:
1. Moved Employee Options help text to be under input (follows new design rules)
2. Added placeholders for grid filters

@LouiseBonnard could you check translation keys/domains?

LouiseBonnard commented:
@sarjon, sure, I'll check it today! Please apply the 'waiting for wording' label when you need me to have a look at keys and domains, I'll check it as soon as possible. ;-)

LouiseBonnard commented on src/Core/Grid/Definition/Factory/EmployeeGridDefinitionFactory.php:
```diff
@@ -140,16 +141,17 @@ protected function getColumns()
-                        ->add(
-                            (new LinkRowAction('edit'))
+                        ->add((new LinkRowAction('edit'))
+                            ->setName($this->trans('Edit', [], 'Admin.Global'))
```
Shouldn't it be localized in _Admin.Actions_ instead?

sarjon replied:
done.

LouiseBonnard commented on src/Core/Grid/Definition/Factory/EmployeeGridDefinitionFactory.php:
```diff
@@ -186,27 +188,39 @@ protected function getFilters()
                 ->setTypeOptions([
                     'required' => false,
+                    'attr' => [
+                        'placeholder' => $this->trans('Search ID', [], 'Admin.Advparameters.Help'),
```
I think it should be localized in _Admin.Actions_ instead.

sarjon replied:
done.

LouiseBonnard commented on src/Core/Grid/Definition/Factory/EmployeeGridDefinitionFactory.php:
```diff
@@ -186,27 +188,39 @@ protected function getFilters()
                 ->setTypeOptions([
                     'required' => false,
+                    'attr' => [
+                        'placeholder' => $this->trans('Search first name', [], 'Admin.Advparameters.Help'),
```
I think it should be localized in _Admin.Actions_ instead.

sarjon replied:
done.

LouiseBonnard commented on src/Core/Grid/Definition/Factory/EmployeeGridDefinitionFactory.php:
```diff
@@ -186,27 +188,39 @@ protected function getFilters()
                 ->setTypeOptions([
                     'required' => false,
+                    'attr' => [
+                        'placeholder' => $this->trans('Search last name', [], 'Admin.Advparameters.Help'),
```
Same here!

sarjon replied:
done.

LouiseBonnard commented on src/Core/Grid/Definition/Factory/EmployeeGridDefinitionFactory.php:
```diff
@@ -186,27 +188,39 @@ protected function getFilters()
                 ->setTypeOptions([
                     'required' => false,
+                    'attr' => [
+                        'placeholder' => $this->trans('Search email', [], 'Admin.Advparameters.Help'),
```
... and here!

sarjon replied:
done.

LouiseBonnard commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/EmployeeController.php:
```diff
@@ -71,59 +84,186 @@ public function indexAction(Request $request, EmployeeFilters $filters)
+                EmployeeCannotChangeItselfException::CANNOT_CHANGE_STATUS => $this->trans('You cannot disable or delete your own account.', 'Admin.Advparameters.Notification'),
+            ],
+            CannotDeleteEmployeeException::class => $this->trans(
+                'Can\'t delete #%id%',
```
_Can't delete #%id%_

sarjon replied:
`\` is used to escape `'` in PHP, so the actual string is: Can't delete #%id%

sarjon committed:
update wording

matks commented:
Thanks @sarjon for handling QA, UX and wording feedbacks 😉 so back to "waiting for QA"

@LouiseBonnard Wording has been updated

matks commented:
Thanks @sarjon
